lib/modules/apostrophe-schemas/index.js

```diff
@@ -977,14 +977,21 @@ module.exports = {
       converters: {
         string: function(req, data, name, object, field, callback) {
           object[name] = self.apos.launder.string(data[name], field.def);
-          if (field.min && (object[name].length < field.min)) {
+          if (object[name] && field.min && (object[name].length < field.min)) {
             // Would be unpleasant, but shouldn't happen since the browser
             // also implements this. We're just checking for naughty scripts
             return callback('min');
           }
-          if (field.max && (object[name].length > field.max)) {
+          // If max is longer than allowed, trim the value down to the max length
+          if (object[name] && field.max && (object[name].length > field.max)) {
             object[name] = object[name].substr(0, field.max);
           }
+          // If field is required but empty (and client side didn't catch that)
+          // This is new and until now if JS client side failed, then it would
+          // allow the save with empty values -Lars
+          if (field.required && (_.isUndefined(data[name]) || !data[name].toString().length)) {
```

Comment:
object[name] is constrained to be a string at this point so you could simplify your code by looking at that rather than data[name].

Reply:
So I thought that also and right now I cannot remember why I did not stay on that exactly. The tests were failing with the prior if so I changed it to this and it turned out to be stable (sorry many hours, no sleep).

One thing that I'm pretty sure about though. I cannot rely on launder because it's too opiniated and changes the value. It will change min max etc. I need to validate / check on what the user input, not what Apostrophe made it to be after laundering.

Reply:
Also def is a problem with launder in relation to validations or resetting to null because the value becomes `3` if default is `3` but the user submitted `null` and that I cannot work with safely.

Reply:
Maybe I did not see the entire context here and if we are before or after launder. I can review this again, but I went around this more than once when doing these changes.